MSAL/src/requests/MSALSilentRequest.m

```diff
@@ -142,25 +144,150 @@ - (void)acquireTokenImpl:(MSALCompletionBlock)completionBlock
+    NSString *familyId = appMetadata ? appMetadata.familyId : @"1";
+    if (![NSString msidIsStringNilOrBlank:familyId])
+    {
+        [self tryFRT:familyId appMetadata:appMetadata completionBlock:completionBlock];
+    }
+    else
+    {
+        [self tryMRRT:completionBlock];
+    }
+}
 
-    self.refreshToken = [self.tokenCache getRefreshTokenWithAccount:_parameters.account.lookupAccountIdentifier
-                                                           familyId:nil
-                                                      configuration:msidConfiguration
-                                                            context:_parameters
-                                                              error:&msidError];
+- (MSIDRefreshToken *)getRefreshToken:(NSString *)familyId error:(NSError **)error
+{
+    _familyId = familyId;
+    MSIDConfiguration *msidConfiguration = _parameters.msidConfiguration;
+    return [self.tokenCache getRefreshTokenWithAccount:_parameters.account.lookupAccountIdentifier
+                                              familyId:familyId
+                                         configuration:msidConfiguration
+                                               context:_parameters
+                                                 error:error];
+}
 
+- (void)tryFRT:(NSString *)familyId appMetadata:(MSIDAppMetadataCacheItem *)appMetadata completionBlock:(MSALCompletionBlock)completionBlock
+{
+    NSError *msidError = nil;
+    _familyRefreshToken = [self getRefreshToken:familyId error:&msidError];
     if (msidError)
     {
         completionBlock(nil, msidError);
         return;
     }
 
-    CHECK_ERROR_COMPLETION(self.refreshToken, _parameters, MSALErrorInteractionRequired, @"No token matching arguments found in the cache")
+    //If no FRT found, try looking for app specific RT
+    if (!_familyRefreshToken)
+    {
+        [self tryMRRT:completionBlock];
+        return;
+    }
 
+    [self refreshAccessToken:_familyRefreshToken completionBlock:^(MSALResult *result, NSError *error)
+     {
+         if (error)
+         {
+             if([self isErrorRecoverableByUserInteraction:error])
+             {
+                 //Udpate app metadata  by resetting familyId if server returns client_mismatch
+                 NSError *msidError = nil;
+                 [self updateAppMetadata:appMetadata
+                             serverError:error
+                              cacheError:&msidError
+                         completionBlock:completionBlock];
+                 
+                 if (msidError)
+                 {
+                     completionBlock(nil, msidError);
+                     return;
+                 }
+                 
+                 //If server returns invalid grant on FRT, look for app specific RT
+                 _refreshToken = [self getRefreshToken:nil error:&msidError];
```

Comment:
nit: consider doing this logic in tryMRRT if familyRefreshToken is set instead, since tryMRRT also calls getRefreshToken, and you could avoid if(! _refreshToken) check there then